### ISSUE TITLE
Allow core self-updates in the UI

### ISSUE THREAD

**Issue by quicksketch** [backdrop/backdrop-issues#3271]
## Describe your issue or idea

In https://github.com/backdrop/backdrop-issues/issues/3105, we added the ability to update Backdrop core via the UI, but we disabled it in the UI by default. It can be re-enabled by manually modifying `installer.settings.json` in the active config directory. We should validate core update functionality once 1.11.0 is out and then consider enabling this option in the UI for a future release of Backdrop.

This is a sub-issue of the main automatic-updates issue at #2018.

<img width="795" alt="screen shot 2018-12-06 at 9 10 13 pm" src="https://user-images.githubusercontent.com/397895/49628835-61bfb180-f99b-11e8-8b4c-028ac7e8441a.png">

---
~~merged PR: https://github.com/backdrop/backdrop/pull/2407~~
Follow up PR: https://github.com/backdrop/backdrop/pull/2440

**Comment by quicksketch** [comment]
@serundeputy has included this task as part of https://github.com/backdrop/backdrop-issues/issues/414. We might separate that portion out and do it here separately.

**Comment by jenlampton** [comment]
Since 414 is for making updates automatic and I don't want to abandon that idea just yet, I think we should add the UI here in this issue. We can revisit automatic updates after the manual ones have been working (safely) for a year or so.

**Comment by jenlampton** [comment]
I'm having a hard time following the code in https://github.com/backdrop/backdrop/pull/2289/files. Where is the core "Update" button supposed to appear? And what is the meaning of the new `core_update` config value? Is that supposed to be for automatic, manual, or to enable the `core "Update" button` UI I haven't located yet? I would love some advice / diection @serundeputy if you get a chance :)

**Comment by serundeputy** [comment]
@jenlampton  Currently there is no UI for core update in `1.11.x`; you have to manually (in code) set `"core_update": 1` in `update.settings.json`. 

Then you  can update at: `/admin/modules/update` by checking the box for Backdrop core and proceeding like any  other module update.

![bcmsup](https://user-images.githubusercontent.com/6763671/49618836-0afcac80-f988-11e8-9c1e-0aa8b20e6fec.png)

**Comment by jenlampton** [comment]
> I was under the impression that both would be made available in 1.12

I don't think it's a good idea to allow automatic updates if we don't even have any proof that ANY updates will even work!  I think we need *lots* of proof. At least a years worth, with a change this dangerous.

**Comment by klonos** [comment]
...yes but the sooner we expose the UI to people, the more testing and feedback we'll get sooner.

**Comment by jenlampton** [comment]
...yes, that's why I'm working on this issue :) So we can get a manual update button, so that we get get some proof that updating works!

**Comment by jenlampton** [comment]
Once we know updating works, we can add the option to make the updates automatic. 

We can't add an automatic option right away, or people will use it. And if updating doesn't work for some reason, they might automatically delete their entire site, (perhaps without even noticing for days or even weeks) and that would be bad. 

We want people to delete their entire sites manually, because then at least they'll know what happened :)

**Comment by jenlampton** [comment]
Okay...  I put a screenshot of the current UI from the PR on #414 into the top issue at https://github.com/backdrop/backdrop-issues/issues/414. The Automatic bit is there, but it's disabled. So not as dangerous as I thought :)

I still think it would be better to wait and put it in enabled, after we're sure it works how we think it's going to work. It's much easier to add things than it is to change things with our BC policy.

**Comment by jenlampton** [comment]
I've [created a PR](https://github.com/backdrop/backdrop/pull/2407) if people want to take this for a spin. 

I've added two new checkboxes, one of which is disabled: 
<img width="795" alt="screen shot 2018-12-06 at 9 10 13 pm" src="https://user-images.githubusercontent.com/397895/49628835-61bfb180-f99b-11e8-8b4c-028ac7e8441a.png">

I've also done a massive cleanup on the rest of the update settings form, including:
* Added fieldsets for "Update checking" "Update notifications" and "Self-Updates".
* Added #descriptions for the fieldsets (pulled text from form element descriptions)
* Used `#states` to properly hide and show fields that are not necessary or don't make sense
* Made all language more human
* Made all radio options into complete sentences
* Replaced "your site" with `this website` in all instances

<img width="807" alt="screen shot 2018-12-06 at 9 15 45 pm" src="https://user-images.githubusercontent.com/397895/49628997-22de2b80-f99c-11e8-949a-e9caeccadc0e.png">

**Comment by klonos** [comment]
> Made all radio options into complete sentences

👍 for that, but I don't think that we are using periods at the end of checkbox/radio labels elsewhere.

**Comment by klonos** [comment]
...wondering if it would be a safer option to be renaming the old core and then extracting the update in its place. We'd then let the end user manually delete the backup of the old core.

If they don't delete the old backups, worse that can happen is to run out of disk space (and given the size of core, that'd be rare). That's way safer than destructively deleting things.

**Comment by klonos** [comment]
This looks great @jenlampton!!! 👍

**Comment by herbdool** [comment]
@klonos can you move your last comment to the related issue? Let's keep this focused on the UI.

**Comment by klonos** [comment]
You are right @herbdool ...I have cross-posted my comment in #414 instead of moving it there, as I feel it still has value here, because the help text we are to introduce says "Allow this website to first **delete** its core, and then replace its core...". This help text will also need to be updated if we do change this to rename + manual delete.

**Comment by jenlampton** [comment]
Just to follow up on the order of operations we covered in today's weekly meeting:
1) merge this as-is
2) fix the update process to include a backup - and update this text in that PR
3) add the backup file to the files_managed table so they can be deleted via the UI
4) maybe also add a setting so that backups are or are-not created for core and/or contrib

**Comment by klonos** [comment]
This is mostly RTBC from me. I have left some comments in the PR for your consideration @jenlampton ...mainly user-facing string changes (labels and help text).

Perhaps also:

- Move the "Self-updates" section above the "Update notifications" section.
- Make "coming soon!" a link either to https://github.com/backdrop/backdrop-issues/issues/414, or to https://backdropcms.org/roadmap#1.x

**Comment by olafgrabienski** [comment]
@jenlampton Great! As a detail, I noticed that the link in the *Self Updates* section which leads to https://backdropcms.org/file-permissions-and-ownership, results in "Access denied".

**Comment by klonos** [comment]
@olafgrabienski that's because that page is unpublished 😉

**Comment by jenlampton** [comment]
I'm not sure about the link...  everything else 👍

**Comment by olafgrabienski** [comment]
> that's because that page is unpublished

Can it be published, or is there an issue with the page?

**Comment by klonos** [comment]
>>that's because that page is unpublished

>Can it be published, or is there an issue regarding the page?

Yes it can 😄 ...this specific page seems to have been authored by @quicksketch back in August, so best person to ask why it is not published yet is him.

**Comment by stpaultim** [comment]
I am having trouble testing the manual update in the UI. I've been posting my comments here: https://github.com/backdrop/backdrop-issues/issues/3433

**Comment by stpaultim** [comment]
Details of my experiments are in #3433 
@serundeputy suggested `looks like Jen used false in the update hook and I think the core code expects 0/1 so maybe this is why it doesn't work??? I can't deep dive right now ... you could test this theory by manually setting the core_update value to 1 in the installer.settings.json config file`

I went through my process again, this time watching the value of core_update in the installer.settings.json file. 

Initially, it is set to `false` before I apply the patch
* If I apply the patch (but do not do a dabase update) and turn on manual updates, it changes to `1`
* If I do a database update before I turn on manual updates, it changes to `true`
* I also tried changing it from `false` to `1` directly in the json file.
I have tried doing the update in each of these scenarios without success.

**Comment by stpaultim** [comment]
Probably my last try for today. I spun up a site at v 1.11.0 and set the permissions as recommended. Then I:
* changed the core_update value to `1` (as recommended by @serundeputy)
* At this point, I do not have option for manual updates
* I applied the patch 2407.patch
* Now, manual update is selected by default on admin/reports/updates/settings
* I do not run database update (which would change core_update value to `true`)
* I try to update core and get the `The website encountered an unexpected error. Please try again later.` message.

**Comment by stpaultim** [comment]
My problem has nothing to do with this patch, because I finally went back and tried the original process of simply editing the installer.settings.json settings (on a 1.11.0 site) to include core_update `1` and it is still not working for me. So, there is something else going on.

**Comment by stpaultim** [comment]
I tried both at the urging of @serundeputy

**Comment by jenlampton** [comment]
@stpaultim when you see the error `The website encountered an unexpected error. Please try again later.` what appears in your log? It would be helpful to know what the problem is at that moment.

**Comment by jenlampton** [comment]
It looks like there may be a bug with manual core update that will hold this issue up. I don't know what the problem is yet, but I'm waiting on an issue from @serundeputy explaining the problem, since it looks like he may already have a solution!

**Comment by herbdool** [comment]
@quinnanya did you run update.php right after applying the PR? I expect it would still be the same result, but difference would be that the checkbox would already be checked (or should be at least).

**Comment by quinnanya** [comment]
@herbdool Nope, didn't run update.php until after the new version of Backdrop core was downloaded via the UI.

**Comment by herbdool** [comment]
Based on @quinnanya and @olafgrabienski feedback I think we could mark this as RTBC. I too had successfully trested the update process before. This PR only makes a minor change and it also seems to work

**Comment by quicksketch** [comment]
I thought we needed to set the default value in `installer.settings.json`, but I forgot we already has a value set there! So with that in place I've merged https://github.com/backdrop/backdrop/pull/2407 into 1.x for 1.12.0. Thanks @jenlampton, @serundeputy, @klonos, @herbdool, @olafgrabienski, @stpaultim, and @quinnanya!

_However_, after merging I realized that the default value in `installer.settings.json` is `false` but the the value we set in the update hook is `true`. We need to decide of core updates are enabled by default and then update both places to be the same.

I'm leaving this needs work for the minor follow-up. If this takes any significant work we need to make a new issue.

**Comment by stpaultim** [comment]
I'm assuming that you meant to reopen this issue, since it requires follow up work?

**Comment by herbdool** [comment]
Going by the comments on the PR it seems that preference might be to leave it as FALSE for now.

**Comment by jenlampton** [comment]
> it seems that preference might be to leave it as FALSE for now.

If the preference is to leave it false, then we won't need the update hook at all. I've filed a new PR to remove it. https://github.com/backdrop/backdrop/pull/2440

**Comment by herbdool** [comment]
Looks good

**Comment by quicksketch** [comment]
Great, merged the follow-up.
